lib/events.js

```diff
@@ -621,22 +629,46 @@ function unwrapListeners(arr) {
   return ret;
 }
 
-function once(emitter, name) {
+function once(emitter, name, options = {}) {
+  const { signal } = { ...options };
+  if (signal !== undefined &&
+      (signal === null ||
+       typeof signal !== 'object' ||
+       !('aborted' in signal))) {
+    return Promise.reject(
+      new ERR_INVALID_ARG_TYPE('options.signal', 'AbortSignal', signal));
+  }
+  if (signal && signal.aborted) {
+    return Promise.reject(
+      lazyDOMException('The operation was aborted', 'AbortError'));
+  }
   return new Promise((resolve, reject) => {
     const errorListener = (err) => {
       emitter.removeListener(name, resolver);
+      if (signal != null)
+        signal.removeEventListener('abort', abortListener);
       reject(err);
     };
     const resolver = (...args) => {
       if (typeof emitter.removeListener === 'function') {
         emitter.removeListener('error', errorListener);
       }
+      if (signal != null)
+        signal.removeEventListener('abort', abortListener);
       resolve(args);
     };
     eventTargetAgnosticAddListener(emitter, name, resolver, { once: true });
     if (name !== 'error') {
       addErrorHandlerIfEventEmitter(emitter, errorListener, { once: true });
     }
+    function abortListener() {
+      emitter.removeListener(name, resolver);
```

Comment:
Shouldn't this use `eventTargetAgnosticRemoveListener`?

Reply:
Ah, right, good catch